File: neo_wake_android/android/src/main/kotlin/xyz/neosapien/neo_wake/WakeCodecPipeline.kt
package xyz.neosapien.neo_wake

/**
 * U3's whole boundary: given a stream of raw seam frames (BLE fragment —
 * `[seq u16][reserved][+cmd flag on Neo2]` header, then Opus or PCM8
 * payload), produce the same score/fire timeline [WakeSpotter] produces for
 * the equivalent Dart chain. Ports `wake_word_service.dart`'s
 * `_onFragment`/`_runHeaderProbe`/`_feedEngine` trio into one class — the
 * pure header-probe ([WakeHeaderProbe]) and framer ([WakePcmFramer]) pieces
 * stay their own small types for the same testability reason the Dart file
 * keeps them `@visibleForTesting` top-level, not nested.
 *
 * No neo_ble wiring here (U5), no command-clip capture (U7), no Dart facade
 * (U6) — this takes bytes in and hands [WakeSpotterStep]s out.
 */
class WakeCodecPipeline(
    private val spotter: WakeSpotter,
    private val codec: NeoWakeAudioCodec,
    private val sampleRate: Int = 16000,
    private val samplesPerFrame: Int = 320, // 20 ms @ 16 kHz mono, one BLE fragment
    headerLenOverride: Int = 0, // 0 = probe at runtime
    probeFramesNeeded: Int = 100,
    private val decoderFactory: () -> OpusDecoding,
) {
    private val framer = WakePcmFramer(WakeSpotter.ADVANCE_SAMPLES)

    private var headerLen = headerLenOverride
    private var probe: WakeHeaderProbe? = if (headerLenOverride == 0) WakeHeaderProbe(probeFramesNeeded) else null
    private val probeDecoders = mutableMapOf<Int, OpusDecoding>()
    private val decoderLazy: Lazy<OpusDecoding> = lazy { decoderFactory() }
    private val decoder: OpusDecoding by decoderLazy

    /** Resolved header length once probing finishes (3 or 4), null while
     * still probing or on an override. */
    var resolvedHeaderLen: Int? = null
        private set

    // Counters mirroring the Dart service's health-tick fields — pure
    // state, no telemetry emission here (that is U5/wiring's job).
    var framesIn = 0
        private set
    var tooShortForHeader = 0
        private set
    var decodeFailed = 0
        private set

    /**
     * Feeds one raw BLE audio fragment through header resolution, decode,
     * re-framing to the chain's 1280-sample advance, and the spotter.
     * Returns zero or more [WakeSpotterStep]s — usually zero (still
     * buffering fragments toward a full advance) or one.
     *
     * [timestampMs] is accepted but not yet consumed — reserved for U5/U7
     * telemetry (matches [NeoWakeAudioFrame.timestampMs]'s same forward
     * surface); do not remove it to silence the unused-parameter warning.
     */
    @Suppress("UNUSED_PARAMETER")
    fun onFragment(raw: ByteArray, timestampMs: Long = 0): List<WakeSpotterStep> {
        framesIn++

        val header = if (headerLen == 0) 4 else headerLen
        if (raw.size <= header) {
            tooShortForHeader++
            breakStream()
            return emptyList()
        }

        if (headerLen == 0) {
            // ponytail: pcm8 has no header-length probe -- the probe is
            // opus-only (it always attempts an Opus decode, see
            // runHeaderProbe's doc comment). U5 wires the negotiated codec +
            // header length in from neo_ble's `cachedCodec`; until then pcm8
            // only functions with an explicit headerLenOverride, never a
            // guessed one.
            if (codec == NeoWakeAudioCodec.PCM8) {
                decodeFailed++
                breakStream()
                return emptyList()
            }
            runHeaderProbe(raw)
            return emptyList()
        }

        val payload = raw.copyOfRange(headerLen, raw.size)
        val samples: ShortArray? = when (codec) {
            NeoWakeAudioCodec.PCM8 -> decodePcm8(payload)
            NeoWakeAudioCodec.OPUS, NeoWakeAudioCodec.UNKNOWN -> decoder.decode(payload, samplesPerFrame)
        }

        if (samples == null) {
            decodeFailed++
            breakStream()
            return emptyList()
        }

        return feedEngine(samples)
    }

    /** A drop is a discontinuity for the chain behind it, not a free slot —
     * see [WakeSpotter]'s file header (R18/KTD13). Every path that skips a
     * fragment routes through here. */
    private fun breakStream() {
        spotter.onFrameDropped()
    }

    private fun feedEngine(samples: ShortArray): List<WakeSpotterStep> {
        val results = mutableListOf<WakeSpotterStep>()
        framer.add(samples) { frame -> results.add(spotter.process(frame)) }
        return results
    }

    /**
     * While probing, decode the SAME fragment at both candidate offsets and
     * let the measurement pick — mirrors `wake_word_service.dart`'s
     * `_runHeaderProbe`, which always attempts an Opus decode: the Dart
     * reference has no pcm8 branch at all. Opus-only — [onFragment] routes
     * pcm8 away from this function before it is ever called (see its
     * `ponytail:` comment); guessing a header length for pcm8 would
     * byte-misalign every sample instead of just failing to fire.
     */
    private fun runHeaderProbe(frame: ByteArray) {
        assert(codec != NeoWakeAudioCodec.PCM8) { "pcm8 must never enter the opus-only header probe" }
        val probe = probe ?: return
        for (len in intArrayOf(3, 4)) {
            if (frame.size <= len) continue
            val dec = probeDecoders.getOrPut(len) { decoderFactory() }
            val payload = frame.copyOfRange(len, frame.size)
            val samples = dec.decode(payload, samplesPerFrame)
            probe.record(len, decoded = samples != null, rms = samples?.let { rmsOf(it) } ?: 0.0)
        }

        if (!probe.done) return

        disposeProbeDecoders()
        headerLen = probe.verdict(3)
        resolvedHeaderLen = headerLen
        this.probe = null
    }

    /** Closes every probe decoder's native handle (they are otherwise only
     * freed via [NeoOpusDecoder.finalize], i.e. whenever GC gets to it) and
     * clears the map. Fakes used in tests don't implement [AutoCloseable],
     * hence the safe cast. */
    private fun disposeProbeDecoders() {
        for (dec in probeDecoders.values) {
            (dec as? AutoCloseable)?.close()
        }
        probeDecoders.clear()
    }

    /** Frees the main lazy decoder's native handle plus any still-live probe
     * decoders, mirroring `wake_word_service.dart`'s `dispose()` /
     * `_disposeProbeDecoders()`. No neo_ble wiring calls this yet (U5's
     * job); exposed here so a disarm/disconnect path can. */
    // ponytail: U5 wires this into the disarm/disconnect path.
    fun close() {
        disposeProbeDecoders()
        // Guard against `by lazy` forcing initialization just to close
        // something that was never created (pcm8-only streams never touch
        // `decoder`).
        if (decoderLazy.isInitialized()) {
            (decoder as? AutoCloseable)?.close()
        }
    }
}

File: neo_wake_ios/ios/Classes/WakeCodecPipeline.swift
import Foundation

/// U3's whole boundary: given a stream of raw seam frames (BLE fragment —
/// `[seq u16][reserved][+cmd flag on Neo2]` header, then Opus or PCM8
/// payload), produce the same score/fire timeline `WakeSpotter` produces for
/// the equivalent Dart chain. Ports `wake_word_service.dart`'s
/// `_onFragment`/`_runHeaderProbe`/`_feedEngine` trio into one class — the
/// pure header-probe (`WakeHeaderProbe`) and framer (`WakePcmFramer`) pieces
/// stay their own small types for the same testability reason the Dart file
/// keeps them `@visibleForTesting` top-level, not nested.
///
/// No neo_ble wiring here (U5), no command-clip capture (U7), no Dart facade
/// (U6) — this takes bytes in and hands `WakeSpotterStep`s out.
public final class WakeCodecPipeline {
    public init(
        spotter: WakeSpotter,
        codec: NeoWakeAudioCodec,
        sampleRate: Int = 16000,
        // Upper bound handed to the Opus decoder as its output-buffer
        // capacity, NOT the actual fragment size (mirrors
        // `wake_word_service.dart`'s `_kSamplesPerFrame` exactly, including
        // its own comment: "it only sizes the output buffer, and a generous
        // bound costs nothing while a tight one truncates"). A real BLE
        // fragment is 10 ms / 160 samples; `OpusDecoding.decode` returns
        // however many samples the decoder actually produced, trimmed to
        // that length, not padded/truncated to this bound — see
        // `NeoOpusDecoder.decode`'s doc comment.
        samplesPerFrame: Int = 320,
        headerLenOverride: Int = 0, // 0 = probe at runtime
        probeFramesNeeded: Int = 100,
        decoderFactory: @escaping () -> OpusDecoding
    ) {
        self.spotter = spotter
        self.codec = codec
        self.sampleRate = sampleRate
        self.samplesPerFrame = samplesPerFrame
        self.headerLen = headerLenOverride
        self.decoderFactory = decoderFactory
        self.framer = WakePcmFramer(frameLength: WakeSpotter.advanceSamples)
        if headerLenOverride == 0 {
            self.probe = WakeHeaderProbe(framesNeeded: probeFramesNeeded)
        }
    }

    private let spotter: WakeSpotter
    private let codec: NeoWakeAudioCodec
    private let sampleRate: Int
    private let samplesPerFrame: Int
    private let decoderFactory: () -> OpusDecoding
    private let framer: WakePcmFramer

    private var headerLen: Int
    private var probe: WakeHeaderProbe?
    private var probeDecoders: [Int: OpusDecoding] = [:]
    private lazy var decoder: OpusDecoding = decoderFactory()

    /// Resolved header length once probing finishes (3 or 4), nil while
    /// still probing or on an override.
    public private(set) var resolvedHeaderLen: Int?

    /// Counters mirroring the Dart service's health-tick fields — pure state,
    /// no telemetry emission here (that is U5/wiring's job).
    public private(set) var framesIn = 0
    public private(set) var tooShortForHeader = 0
    public private(set) var decodeFailed = 0

    /// Feeds one raw BLE audio fragment through header resolution, decode,
    /// re-framing to the chain's 1280-sample advance, and the spotter.
    /// Returns zero or more `WakeSpotterStep`s — usually zero (still
    /// buffering fragments toward a full advance) or one.
    @discardableResult
    public func onFragment(_ raw: [UInt8], timestampMs: Int64 = 0) throws -> [WakeSpotterStep] {
        framesIn += 1

        let header = headerLen == 0 ? 4 : headerLen
        guard raw.count > header else {
            tooShortForHeader += 1
            breakStream()
            return []
        }

        if headerLen == 0 {
            // ponytail: pcm8 has no header-length probe -- the probe is
            // opus-only (it always attempts an Opus decode, see
            // runHeaderProbe's doc comment). U5 wires the negotiated codec +
            // header length in from neo_ble's `cachedCodec`; until then pcm8
            // only functions with an explicit `headerLenOverride`, never a
            // guessed one.
            guard codec != .pcm8 else {
                decodeFailed += 1
                breakStream()
                return []
            }
            runHeaderProbe(raw)
            return []
        }

        let payload = Array(raw[headerLen...])
        let samples: [Int16]?
        switch codec {
        case .pcm8:
            samples = decodePcm8(payload)
        case .opus, .unknown:
            samples = decoder.decode(payload, frameSize: samplesPerFrame)
        }

        guard let pcm = samples else {
            decodeFailed += 1
            breakStream()
            return []
        }

        return try feedEngine(pcm)
    }

    /// A drop is a discontinuity for the chain behind it, not a free slot —
    /// see `WakeSpotter`'s file header (R18/KTD13). Every path that skips a
    /// fragment routes through here.
    private func breakStream() {
        spotter.onFrameDropped()
    }

    // Post-fire cooldown (mirrors the Dart service's _kLockoutMs). Without it
    // the embedding ring keeps scoring the same "Neo SimSim" every 80 ms step
    // for ~1.3 s, so one spoken wake word fires dozens of times. On a real fire
    // we clear the ring (onDetection) AND suppress further fires for the lockout
    // window, so a single utterance = a single fire and the open/close toggle
    // stays sane.
    private var lockoutUntilMs: Double = 0
    private static let lockoutMs: Double = 1500

    private func feedEngine(_ samples: [Int16]) throws -> [WakeSpotterStep] {
        var results: [WakeSpotterStep] = []
        var thrown: Error?
        framer.add(samples) { frame in
            guard thrown == nil else { return }
            do {
                let step = try spotter.process(frame)
                if step.fired {
                    let now = Date().timeIntervalSince1970 * 1000
                    if now < lockoutUntilMs {
                        // Within the cooldown — a re-fire on the still-decaying
                        // ring. Surface the step but not as a fire.
                        results.append(WakeSpotterStep(stepIndex: step.stepIndex, score: step.score, fired: false))
                    } else {
                        results.append(step)
                        spotter.onDetection() // clear the ring so the same word can't re-fire
                        lockoutUntilMs = now + WakeCodecPipeline.lockoutMs
                    }
                } else {
                    results.append(step)
                }
            } catch {
                thrown = error
            }
        }
        if let thrown { throw thrown }
        return results
    }

    /// While probing, decode the SAME fragment at both candidate offsets and
    /// let the measurement pick — mirrors `wake_word_service.dart`'s
    /// `_runHeaderProbe`, which always attempts an Opus decode: the Dart
    /// reference has no pcm8 branch at all. Opus-only — `onFragment` routes
    /// pcm8 away from this function before it is ever called (see its
    /// `ponytail:` comment); guessing a header length for pcm8 would
    /// byte-misalign every sample instead of just failing to fire.
    private func runHeaderProbe(_ frame: [UInt8]) {
        assert(codec != .pcm8, "pcm8 must never enter the opus-only header probe")
        guard let probe else { return }
        for len in [3, 4] {
            guard frame.count > len else { continue }
            let dec = probeDecoders[len] ?? decoderFactory()
            probeDecoders[len] = dec
            let payload = Array(frame[len...])
            let samples = dec.decode(payload, frameSize: samplesPerFrame)
            probe.record(headerLen: len, decoded: samples != nil, rms: samples.map(rmsOf) ?? 0.0)
        }

        guard probe.done else { return }

        probeDecoders.removeAll()
        headerLen = probe.verdict(fallback: 3)
        resolvedHeaderLen = headerLen
        self.probe = nil
    }
}
